e2e/disconnectedclients/disconnectedclients_test.go

```diff
@@ -0,0 +1,128 @@
+package disconnectedclients
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/hashicorp/nomad/e2e/e2eutil"
+	"github.com/hashicorp/nomad/helper/uuid"
+	"github.com/hashicorp/nomad/testutil"
+	"github.com/stretchr/testify/require"
+)
+
+const ns = ""
+
+// typical wait times for this test package
+var wait30s = &e2eutil.WaitConfig{Interval: time.Second, Retries: 30}
+var wait60s = &e2eutil.WaitConfig{Interval: time.Second, Retries: 60}
+
+func TestDisconnectedClients(t *testing.T) {
+	nomad := e2eutil.NomadClient(t)
+
+	e2eutil.WaitForLeader(t, nomad)
+	e2eutil.WaitForNodesReady(t, nomad, 2) // needs at least 2 to test replacement
+
+	t.Run("AllocReplacementOnShutdown", testDisconnected_AllocReplacementOnShutdown)
+}
+
+// disconnectedClientsCleanup sets up a cleanup function to make sure
+// we've waited for all the nodes to come back up between tests
+func disconnectedClientsCleanup(t *testing.T) func() {
+	nodeIDs := []string{}
+	nodeStatuses, err := e2eutil.NodeStatusList()
+	require.NoError(t, err)
+	for _, nodeStatus := range nodeStatuses {
+		nodeIDs = append(nodeIDs, nodeStatus["ID"])
+	}
+	return func() {
+		nomad := e2eutil.NomadClient(t)
+		e2eutil.WaitForNodesReady(t, nomad, len(nodeIDs))
+	}
+}
+
+// testDisconnected_AllocReplacementOnShutdown tests that allocations on
+// clients that are shut down and marked disconnected are replaced
+func testDisconnected_AllocReplacementOnShutdown(t *testing.T) {
+
+	jobIDs := []string{}
+	t.Cleanup(disconnectedClientsCleanup(t))
+	t.Cleanup(e2eutil.CleanupJobsAndGC(t, &jobIDs))
+
+	jobID := "test-lost-allocs-" + uuid.Short()
+
+	err := e2eutil.Register(jobID, "./input/lost_simple.nomad")
+	require.NoError(t, err)
+	jobIDs = append(jobIDs, jobID)
+
+	err = e2eutil.WaitForAllocStatusExpected(jobID, ns,
+		[]string{"running", "running"})
+	require.NoError(t, err, "job should be running")
+
+	// pick a node to make our lost node
+	allocs, err := e2eutil.AllocsForJob(jobID, ns)
+	require.NoError(t, err, "could not query allocs for job")
+	require.Len(t, allocs, 2, "could not find 2 allocs for job")
+
+	lostAlloc := allocs[0]
+	lostAllocID := lostAlloc["ID"]
+	disconnectedNodeID := lostAlloc["Node ID"]
+	otherAllocID := allocs[0]["ID"]
```

Comment:
Is this supposed to have a different index? lostAlloc and otherAlloc both point to index 0.

Reply:
🤦 We missed that on the original PR! Will fix!